File: 1992-find-all-groups-of-farmland/1992-find-all-groups-of-farmland.py
class Solution:
    
    def bfs(self,land,directions,i,j,ans):
        queue=deque()
        queue.append((i,j))
        land[i][j]=0
    
        while queue:
            er,ec=queue.popleft()
            for dr,dc in directions:
                nr=dr+er
                nc=dc+ec
                if nr>=0 and nr<len(land) and nc>=0 and nc<len(land[0]) and land[nr][nc]==1:
                    land[nr][nc]=0
                    queue.append((nr,nc))
        ans.append([i,j,er,ec])            
                    
    def findFarmland(self, land: List[List[int]]) -> List[List[int]]:
        ans=[]
        m=len(land)
        n=len(land[0])
        stack=[]
        directions=[[-1,0],[0,1],[0,-1],[1,0]]
        for i in range(m):
            for j in range(n):
                if land[i][j]==1:
                    self.bfs(land,directions,i,j,ans)
        return ans            
                    
        

File: 1992-find-all-groups-of-farmland/1992-find-all-groups-of-farmland.cpp
class Solution {
public:
    int dx[5]={-1,0,1,0,-1};
    int m;
    int n;
    vector<vector<int>> ans;
    void bfs(vector<vector<int>>& land,int sr,int sc){
        int dr=-1;
        int dc=-1;
        queue<pair<int,int>> q;
        q.push({sr,sc});
        land[sr][sc]=0;
        while(!q.empty()){
            pair<int,int> t=q.front();
            q.pop();
            dr=t.first;
            dc=t.second;
            for(int i=0;i<4;i++){
                int nr=dr+dx[i];
                int nc=dc+dx[i+1];
                if(nr>=0 and nr<m and nc>=0 and nc<n and land[nr][nc]==1){
                    land[nr][nc]=0;
                    q.push({nr,nc});
                    
                }
            }
        }
        ans.push_back({sr,sc,dr,dc});
    }
    vector<vector<int>> findFarmland(vector<vector<int>>& land) {
        //use bfs approach
        m=land.size();
        n=land[0].size();
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(land[i][j]==1){
                    bfs(land,i,j);
                }
            }
        }
        return ans;
        
    }
};